onnxruntime/core/optimizer/if_to_where_transformer.cc

```diff
@@ -0,0 +1,226 @@
+        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                               "Node input '", input_arg->Name(), "' not found in main graph for node '", node.Name(), "'");
+      }
+      inputs.push_back(mapped_input);
+    }
+
+    // Outputs
+    std::vector<NodeArg*> outputs;
+    for (const auto* output_arg : node.OutputDefs()) {
+      // Create a unique NodeArg in main graph for each output
+      NodeArg* mapped_output = &main_graph.GetOrCreateNodeArg(output_arg->Name(), output_arg->TypeAsProto());
+      name_to_nodearg[output_arg->Name()] = mapped_output;
+      outputs.push_back(mapped_output);
+    }
+
+    // Add node to main graph
+    main_graph.AddNode(
+        main_graph.GenerateNodeName(node.OpType()),
+        node.OpType(),
+        node.Description(),
+        inputs,
+        outputs,
+        &node.GetAttributes(),
+        node.Domain());
+  }
+
+  // Map subgraph outputs to main graph NodeArgs
+  for (const auto* output : subgraph.GetOutputs()) {
+    auto it = name_to_nodearg.find(output->Name());
+    NodeArg* mapped_output = (it != name_to_nodearg.end()) ? it->second : main_graph.GetNodeArg(output->Name());
+    if (!mapped_output) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                             "Missing output: ", output->Name(), " in main graph after inlining.");
+    }
+    subgraph_outputs.push_back(mapped_output);
+  }
+
+  return Status::OK();
+}
+
+static bool CanTransformIfToWhere(const Node& if_node) {
+  const auto& outputs = if_node.OutputDefs();
+  for (const auto* output : outputs) {
+    const auto* type_proto = output->TypeAsProto();
+    if (!type_proto) return false;
+
+    // If any output is optional, we skip the transformation
+    if (type_proto->value_case() == onnx::TypeProto::kOptionalType) {
+      return false;
+    }
+
+    // Optionally, ensure it's a tensor
+    if (type_proto->value_case() != onnx::TypeProto::kTensorType) {
+      return false;
+    }
+  }
+  return true;
+}
+
+Status IfToWhereTransformer::ApplyImpl(Graph& graph,
+                                       bool& modified,
+                                       int graph_level,
+                                       const logging::Logger& logger) const {
+  modified = false;
+  (void)graph_level;
+  for (auto it = graph.Nodes().begin(), end = graph.Nodes().end(); it != end; ++it) {
+    Node& if_node = *it;
+    if (if_node.OpType() != "If")
+      continue;
+
+    if (!graph_utils::IsSupportedProvider(if_node, GetCompatibleExecutionProviders())) {
+      // If not Qnn Execution Provider
+      continue;
+    }
+
+    if (!CanTransformIfToWhere(if_node)) {
+      LOGS(logger, VERBOSE) << "Skipping IfToWhere transformation for node " << if_node.Name()
+                            << " due to optional or unsupported output types.";
+      continue;
+    }
+
+    const NodeArg* cond_arg = if_node.InputDefs()[0];
+    if (!cond_arg)
+      continue;
```

Comment:
`cond` is a required input so model load should have failed if it wasn't present.

It's required to be a `bool` so that check should also be unnecessary.